### PR TITLE
[WIP] Pytorch build: build _C.so via CMake

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

File: tools/cpp_build/libtorch/CMakeLists.txt [hunk 1]
```diff
@@ -163,56 +199,57 @@ add_custom_command(
   WORKING_DIRECTORY "${TORCH_SRC_DIR}/..")
 
 set(TORCH_SRCS
```
Comment (anderspapitto) [inline tools/cpp_build/libtorch/CMakeLists.txt]: note that this list has been sorted but otherwise unchanged

File: setup.py [hunk 2]
```diff
@@ -473,6 +477,38 @@ def run(self):
...
+                '-DNANOPB_BUILD_PATH=' + lib_path,
+                '-DTORCH_DEP_LIB_PATH=' + lib_path,
+                '-DPYTHON_EXECUTABLE:FILEPATH={}'.format(sys.executable),
+                '-DPYTHON_INCLUDE_DIR={}'.format(sysconfig.get_python_inc()),
+                '-DWITH_NUMPY=ON',
+                '-DWITH_DISTRIBUTED=ON',
+                '-DWITH_CUDA=ON',
+                '-DWITH_NCCL=ON',
+                '-DCUDA_LIB_PATH=/usr/local/cuda/lib64',
+                '-G', 'Ninja',
```
Comment (anderspapitto) [inline setup.py]: TODO don't hardcode ninja

File: setup.py [hunk 3]
```diff
@@ -473,6 +477,40 @@ def run(self):
 
             self.copy_file(export_lib, target_lib)
 
+    def build_extension(self, ext):
+        if ext.name == 'torch._C':
+            cmake_build_dir = os.path.join(self.build_lib, 'torch')
```
Comment (anderspapitto) [inline setup.py]: TODO: build somewhere else and copy selected files into this directory?

**Comment by apaszke** [comment]
Why do we want to build the Python binding parts via CMake when setuptools works perfectly fine for that purpose?

File: setup.py (hunk 3 above)
Comment (orionr) [reply setup.py]: Best would be to build in the `build/` directory at the root level, since build_deps seems to use that as well.

File: setup.py [hunk 4]
```diff
@@ -473,6 +479,40 @@ def run(self):
...
+            try:
+                os.makedirs(cmake_build_dir)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise
+
+            cmake_cmd = [
```
Comment (orionr) [inline setup.py]: Any chance we could put this into `tools/build_pytorch_libs.sh` instead? Eventually we want this built as a part of the root-level cmake (and therefore remove build_pytorch_libs.sh) and having it here will make that more difficult.

File: setup.py [hunk 5]
```diff
@@ -473,6 +479,40 @@ def run(self):
...
+                '-DTORCH_DEP_LIB_PATH=' + lib_path,
+                '-DPYTHON_EXECUTABLE:FILEPATH={}'.format(sys.executable),
+                '-DPYTHON_INCLUDE_DIR={}'.format(sysconfig.get_python_inc()),
+                '-DWITH_NUMPY=ON',
+                '-DWITH_DISTRIBUTED=ON',
+                '-DNO_CUDA=' + ('OFF' if WITH_CUDA else 'ON'),
+                '-DWITH_NCCL=ON',
+                '-DCUDA_LIB_PATH=/usr/local/cuda/lib64'
+            ]
+            if WITH_NINJA:
```
Comment (orionr) [inline setup.py]: This will also be taken care of by `build_pytorch_libs.sh`

File: tools/cpp_build/libtorch/CMakeLists.txt [hunk 6]
```diff
@@ -62,7 +66,33 @@ if(NOT NO_CUDA)
...
 
-add_definitions(-DNO_PYTHON -D_FORCE_INLINES)
+if(NOT FULL_BUILD)
+  add_definitions(-DNO_PYTHON)
+endif()
+
+add_definitions(-D_FORCE_INLINES)
+
+if(FULL_BUILD)
+  include_directories(${PYTHON_INCLUDE_DIR})
```
Comment (orionr) [inline tools/cpp_build/libtorch/CMakeLists.txt]: Nice!

**Review by orionr — COMMENTED** [review]
Great start! Let's see if you can move this under build_pytorch_libs.sh

**Pull request closed without merging**